WiFiManager.cpp

```diff
@@ -136,24 +168,44 @@ boolean WiFiManager::autoConnect(char const *apName, char const *apPassword) {
   //String ssid = getSSID();
   //String pass = getPassword();
 
-  // attempt to connect; should it fail, fall back to AP
+  // device will attempt to connect by itself; wait 10 secs
+  // to see if it succeeds and should it fail, fall back to AP
   WiFi.mode(WIFI_STA);
-
-  if (connectWifi("", "") == WL_CONNECTED)   {
-    DEBUG_WM(F("IP Address:"));
-    DEBUG_WM(WiFi.localIP());
-    //connected
-    return true;
-  }
+  unsigned long startedAt = millis();
+    while(millis() - startedAt < 10000)
+    {
+        delay(100);
+        if (WiFi.status()==WL_CONNECTED) {
+			float waited = (millis()- startedAt);
+			DEBUG_WM(F("After waiting "));
+			DEBUG_WM(waited/1000);
+			DEBUG_WM(F(" secs local ip: "));
+			DEBUG_WM(WiFi.localIP());
+            return true;
+		}
+    }
 
   return startConfigPortal(apName, apPassword);
 }
 
+boolean  WiFiManager::startConfigPortal() {
+  String ssid = "ESP" + String(ESP.getChipId());
+  return startConfigPortal(ssid.c_str());
+}
+
 boolean  WiFiManager::startConfigPortal(char const *apName, char const *apPassword) {
   //setup AP
-  WiFi.mode(WIFI_AP_STA);
-  DEBUG_WM("SET AP STA");
-
+  int connRes = WiFi.waitForConnectResult();
+  if (connRes == WL_CONNECTED){
+	  WiFi.mode(WIFI_AP_STA); //Dual mode works fine if it is connected to WiFi
+	  DEBUG_WM("SET AP STA");
+  	}
+  	else {
+    WiFi.mode(WIFI_AP); // Dual mode becomes flaky if not connected to a WiFi network.
```

Comment:
Modified scanning to take place before configuration portal is created for browser interface, JSON interface unchanged. 

This avoids the client device becoming disconnected from the WiFi network during scanning. 

There was also an issue previously that, because scanning took a long time, users would refresh, go back, click on other selections etc in the belief nothing was happening. This would build up a backlog of requests on the web server that it would queue and carry out sequentially but for which the client device was no longer waiting for a response, thus causing confusion.